(RK-340) Prefer SHA256 checksum for verifying modules

Magisus opened the pull request:
FIPS mode does not allow md5 to be used, to this updates r10k to always prefer using SHA256 when verifying modules. If the Forge does not have SHA256 data and FIPS mode is not enabled, it will fall back to verifying with md5. If FIPS mode is enabled, it will error.

Magisus committed:
(RK-340) Use SHA256 when verifying modules in FIPS mode

FIPS mode does not allow the use of md5 hashes, so this commit updates
the module install code to use SHA256 to verify downloads when running
in FIPS mode.

Magisus committed:
(RK-340) Add FIPS mode detection

This commit adds a method to the Platform module for detecting whether
or not the current platform is running in FIPS mode. Currently only
guaranteed to detect FIPS mode on RedHat 7 systems.

Magisus committed:
(RK-340) Prefer SHA256 always, fall back to md5 when not in FIPS mode

This commit updates the logic around checksum verification when
installing a module to always prefer SHA256 and only use md5 if the
Forge data does not contain an SHA256 checksum.

joshcooper commented on lib/r10k/forge/module_release.rb:
```diff
@@ -108,12 +115,71 @@ def download
 
-        md5_of_tarball = Digest::MD5.hexdigest(File.read(@tarball_cache_path, mode: 'rb'))
+        contents = File.read(@tarball_cache_path, mode: 'rb')
+        sha256_of_tarball = Digest::SHA256.hexdigest(contents)
```
One optimization is to call `Digest::SHA256.file(@tarball_cache_path).hexdigest` so that you don't have to read the entire file into memory. It should handle binary mode on Windows.

Magisus replied:
Thanks I'll give that a shot

justinstoller commented on lib/r10k/forge/module_release.rb:
```diff
@@ -108,12 +115,70 @@ def download
+        else
+          begin
+            verify_sha256_from_forge(sha256_of_tarball)
+          rescue R10K::Error
```
This is just using an exception for control flow and should be done differently.

justinstoller commented on lib/r10k/forge/module_release.rb:
```diff
@@ -108,12 +115,70 @@ def download
+            md5_of_tarball = Digest::MD5.file(@tarball_cache_path).hexdigest
+
+            if @md5_file_path.exist?
+              verify_from_md5_file(md5_of_tarball)
```
If we parameterized these methods could we condense them? It seems that they all do largely the same thing but are reading different instance variables (and logging a formulaic, but slightly different message).

Magisus replied:
Yeah I was thinking about that. I'll look into it.

justinstoller commented on lib/r10k/forge/module_release.rb:
```diff
@@ -205,6 +270,13 @@ def cleanup_md5_file_path
       end
+
+      # Remove the module release sha256.
+      def cleanup_sha256_file_path
```
This convention is wild.

Magisus replied:
What do you mean, specifically?

justinstoller replied:
the abstraction over a call to `#delete`, and an existence check that most callers will already know the answer to.

not your code, but i dislike the conventions around the `cleanup_*_path` methods, and prefer methods that are _not_ part of an imperative, high level, public api to only act on their explicit inputs.

Magisus replied:
I can ... cleanup... that too ;)

justinstoller commented on lib/r10k/util/platform.rb:
```diff
@@ -14,6 +16,16 @@ def self.platform
+      # toggled via a file.
+      def self.fips?
+        if File.exist?(FIPS_FILE)
+          File.read(FIPS_FILE).chomp == "1"
```
I'm kinda surprised the r10k user would be able to read this file

justinstoller replied:
I'll be damned, it's world readable:
```
root@l3ieludaqizuhkz ~]# ls -la /proc/sys/crypto/
total 0
dr-xr-xr-x 1 root root 0 Jul 29 20:30 .
dr-xr-xr-x 1 root root 0 Jul 29 20:28 ..
-r--r--r-- 1 root root 0 Jul 29 20:30 fips_enabled
```

Magisus replied:
Maybe it can't? Is the r10k user usually different than the puppet user? Because that's how puppet is doing it... see https://github.com/puppetlabs/puppet/blob/ba4d1a1aba0095d3c70b98fea5c67434a4876a61/lib/puppet/util/platform.rb#L24-L26. (This also worked in my manual testing, but there I was running as root on a pooler machine, which I guess might not be typical?)

Magisus replied:
Works for me :P

Magisus committed:
(RK-340) Refactor verify methods

This commit refactors the verification of module downloads to be less
repetitive.

Magisus commented on lib/r10k/forge/module_release.rb:
```diff
@@ -104,54 +111,79 @@ def download
+      def verify_from_file(tarball_checksum, checksum_file_path)
+        checksum_from_file = File.read(checksum_file_path).strip
+        if tarball_checksum != checksum_from_file
+          logger.error "Checksum of #{@tarball_cache_path} (#{tarball_checksum}) does not match checksum #{checksum_from_file} in #{checksum_file_path}. Removing both files."
```
Do you think it's worth passing in the checksum type here to maintain putting `MD5` vs `SHA256` in the error message?

justinstoller replied:
I don't think so, since a) I can't think of a use case that would involve scripting against checksum mismatches and b) the type of the checksum should still be inferable from the path (eg. `puppetlabs-concat-6.0.0.md5` or however that path works out). So I'd say the risk of breaking folks is very low and there's no loss in data for the end user.

But I'd be curious what the community thinks. @binford2k do you know if folks would be upset if we changed the checksum mismatch error message?

binford2k replied:
I've not ever seen anyone write tooling that depends on this. And I don't see a way to do it without changing the message, so I see no blocker here.
